File: gateway/controller/fee.go
package controller

import (
	"context"
	"fmt"
	"strconv"
	"time"

	feepb "github.com/essayZW/hpcmanager/fee/proto"
	"github.com/essayZW/hpcmanager/gateway/middleware"
	gatewaypb "github.com/essayZW/hpcmanager/gateway/proto"
	"github.com/essayZW/hpcmanager/gateway/request/json"
	"github.com/essayZW/hpcmanager/gateway/response"
	"github.com/essayZW/hpcmanager/gateway/utils"
	"github.com/essayZW/hpcmanager/proto"
	"github.com/gin-gonic/gin"
	"go-micro.dev/v4/client"
)

type fee struct {
	feeService feepb.FeeService
}

// ping /api/fee/ping GET ping测试
func (f *fee) ping(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)
	res, err := f.feeService.Ping(context.Background(), &proto.Empty{
		BaseRequest: baseRequest,
	})
	var resp *response.Response
	if err != nil {
		resp = response.New(500, err, false, "ping fail!")
	} else {
		resp = response.New(200, res, true, "success")
	}
	resp.Send(ctx)
}

// paginationGetNodeDistributeBill /api/fee/distribute GET 分页查询用户节点独占账单
func (f *fee) paginationGetNodeDistributeBill(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	pageIndex, pageSize, err := utils.ParsePagination(ctx)
	if err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.PaginationGetNodeDistributeBill(c, &feepb.PaginationGetNodeDistributeBillRequest{
		BaseRequest: baseRequest,
		PageIndex:   int32(pageIndex),
		PageSize:    int32(pageSize),
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("查询账单信息失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}

	respData := &response.PaginationQueryResponse{
		Data:  resp.Bills,
		Count: int(resp.Count),
	}
	if resp.Bills == nil {
		respData.Data = make([]*feepb.NodeDistributeBill, 0)
	}
	httpResp := response.New(200, respData, true, "success")
	httpResp.Send(ctx)
}

// payNodeDistributeBill /api/fee/distribute PUT 支付机器独占账单
func (f *fee) payNodeDistributeBill(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	var param json.PayNodeDistributeBillParam
	if err := ctx.ShouldBindJSON(&param); err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()
	resp, err := f.feeService.PayNodeDistributeBill(c, &feepb.PayNodeDistributeBillRequest{
		BaseRequest: baseRequest,
		Id:          int32(param.ID),
		PayMoney:    param.PayMoney,
		PayMessage:  param.PayMessage,
		PayType:     int32(param.PayType),
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("支付账单失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}
	if !resp.Success {
		httpResp := response.New(200, nil, false, "账单没有发生任何变化")
		httpResp.Send(ctx)
		return
	}
	httpResp := response.New(200, nil, true, "success")
	httpResp.Send(ctx)
}

// getNodeDistributeFeeRate /api/fee/rate/distribute GET 查询机器节点独占费率
func (f *fee) getNodeDistributeFeeRate(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()
	resp, err := f.feeService.GetNodeDistributeFeeRate(c, &feepb.GetNodeDistributeFeeRateRequest{
		BaseRequest: baseRequest,
	})
	if err != nil {
		httpResp := response.New(200, nil, false, "信息查询失败")
		httpResp.Send(ctx)
		return
	}

	httpResp := response.New(200, map[string]float64{
		"rate36CPU": resp.Rate36CPU,
		"rate4GPU":  resp.Rate4GPU,
		"rate8GPU":  resp.Rate8GPU,
	}, true, "success")
	httpResp.Send(ctx)
}

// paginationGetNodeWeekUsageBills /api/fee/usage/week GET 分页查询机器节点机时周账单
func (f *fee) paginationGetNodeWeekUsageBills(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	pageIndex, pageSize, err := utils.ParsePagination(ctx)
	if err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	startTime, endTime, err := utils.ParseDateRange(ctx)
	if err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.PaginationGetNodeWeekUsageBillRecords(c, &feepb.PaginationGetNodeWeekUsageBillRecordsResquest{
		BaseRequest:   baseRequest,
		PageIndex:     int32(pageIndex),
		PageSize:      int32(pageSize),
		StartTimeUnix: startTime.Unix(),
		EndTimeUnix:   endTime.Unix(),
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("查询账单信息失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}
	responseData := response.PaginationQueryResponse{
		Data:  resp.Bills,
		Count: int(resp.Count),
	}
	if resp.Bills == nil {
		responseData.Data = make([]*feepb.NodeWeekUsageBill, 0)
	}
	httpResp := response.New(200, responseData, true, "success")
	httpResp.Send(ctx)
}

// paginationGetNodeWeekUsageBillsGroupByGroupID 分页查询机时周账单并按照组ID进行分组
func (f *fee) paginationGetNodeWeekUsageBillsGroupByGroupID(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	pageIndex, pageSize, err := utils.ParsePagination(ctx)
	if err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	payFlagStr, ok := ctx.GetQuery("payFlag")
	if !ok {
		httpResp := response.New(200, nil, false, "缺少payFlag参数")
		httpResp.Send(ctx)
		return
	}
	payFlag, err := strconv.ParseBool(payFlagStr)
	if err != nil {
		httpResp := response.New(200, nil, false, "payFlag 必须是一个Bool值")
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.PaginationGetUserGroupUsageBillRecords(c, &feepb.PaginationGetUserGroupUsageBillRecordsRequest{
		BaseRequest: baseRequest,
		PageIndex:   int32(pageIndex),
		PageSize:    int32(pageSize),
		PayFlag:     payFlag,
	})

	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("查询账单失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}

	responseData := response.PaginationQueryResponse{
		Count: int(resp.Count),
		Data:  resp.Bills,
	}
	if resp.Bills == nil {
		responseData.Data = make([]*feepb.NodeWeekUsageBillForUserGroup, 0)
	}
	httpResp := response.New(200, responseData, true, "success")
	httpResp.Send(ctx)
}

// payGroupNodeUsageBill /api/fee/usage/group/bill PUT 支付用户组机器节点机时账单
func (f *fee) payGroupNodeUsageBill(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	var param json.PayGroupNodeUsageBillParam
	if err := ctx.ShouldBindJSON(&param); err != nil {
		httpResp := response.New(200, nil, false, "参数验证失败")
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.PayGroupNodeUsageBill(c, &feepb.PayGroupNodeUsageBillRequest{
		BaseRequest: baseRequest,
		UserGroupID: int32(param.UserGroupID),
		PayType:     int32(param.PayType),
		PayMessage:  param.PayMessage,
		NeedFee:     param.NeedFee,
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("缴费失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}

	httpResp := response.New(200, map[string]int{
		"count": int(resp.PayCount),
	}, true, "success")
	httpResp.Send(ctx)
}

// getNodeUsageFeeRate /api/fee/rate/usage GET 查询机器时长费率信息
func (f *fee) getNodeUsageFeeRate(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.GetNodeUsageFeeRate(c, &feepb.GetNodeUsageFeeRateRequest{
		BaseRequest: baseRequest,
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("查询费率信息失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}

	httpResp := response.New(200, map[string]float64{
		"cpu": resp.Cpu,
		"gpu": resp.Gpu,
	}, true, "success")
	httpResp.Send(ctx)
}

// paginationGetNodeQuotaBills /api/fee/quota GET 分页查询机器存储账单
func (f *fee) paginationGetNodeQuotaBills(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	pageIndex, pageSize, err := utils.ParsePagination(ctx)
	if err != nil {
		httpResp := response.New(200, nil, false, err.Error())
		httpResp.Send(ctx)
		return
	}

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.PaginationGetNodeQuotaBill(c, &feepb.PaginationGetNodeQuotaBillRequest{
		BaseRequest: baseRequest,
		PageIndex:   int32(pageIndex),
		PageSize:    int32(pageSize),
	})
	if err != nil {
		httpResp := response.New(200, nil, false, fmt.Sprintf("查询账单信息失败: %s", err.Error()))
		httpResp.Send(ctx)
		return
	}
	respData := &response.PaginationQueryResponse{
		Data:  resp.Bills,
		Count: int(resp.Count),
	}
	if resp.Bills == nil {
		respData.Data = make([]*feepb.NodeQuotaBill, 0)
	}
	httpResp := response.New(200, respData, true, "success")
	httpResp.Send(ctx)
}

// getNodeQuotaFeeRate /api/fee/rate/quota GET  查询机器存储费率
func (f *fee) getNodeQuotaFeeRate(ctx *gin.Context) {
	baseReq, _ := ctx.Get(middleware.BaseRequestKey)
	baseRequest := baseReq.(*gatewaypb.BaseRequest)

	c, cancel := context.WithTimeout(context.Background(), time.Duration(5)*time.Second)
	defer cancel()

	resp, err := f.feeService.GetNodeQuotaFeeRate(c, &feepb.GetNodeQuotaFeeRateRequest{
		BaseRequest: baseRequest,
	})
	if err != nil {
		httpResp := response.New(200, nil, false, "信息查询失败")
		httpResp.Send(ctx)
		return
	}

	httpResp := response.New(200, map[string]interface{}{
		"basic": resp.Basic,
		"extra": resp.Extra,
	}, true, "success")
	httpResp.Send(ctx)
}

func (f *fee) Registry(router *gin.RouterGroup) *gin.RouterGroup {
	feeRouter := router.Group("/fee")

	feeRouter.GET("/ping", f.ping)
	middleware.RegistryExcludeAPIPath("GET:/api/fee/ping")

	feeRouter.GET("/distribute", f.paginationGetNodeDistributeBill)
	feeRouter.PUT("/distribute", f.payNodeDistributeBill)

	feeRouter.GET("/rate/distribute", f.getNodeDistributeFeeRate)
	feeRouter.GET("/rate/usage", f.getNodeUsageFeeRate)
	feeRouter.GET("/rate/quota", f.getNodeQuotaFeeRate)

	feeRouter.GET("/usage/week", f.paginationGetNodeWeekUsageBills)
	feeRouter.GET("/usage/group/week", f.paginationGetNodeWeekUsageBillsGroupByGroupID)
	feeRouter.PUT("/usage/group/bill", f.payGroupNodeUsageBill)

	feeRouter.GET("/quota", f.paginationGetNodeQuotaBills)
	return feeRouter
}

func NewFee(client client.Client) Controller {
	feeService := feepb.NewFeeService("fee", client)
	return &fee{
		feeService: feeService,
	}
}

File: web/src/api/fee.ts
import { undefinedWithDefault } from '../utils/obj';
import {
  ApiBaseURL,
  ApiRequest,
  PaginationQueryResponse,
  PingResponse,
} from './api';

// 费用服务ping测试
export async function ping(): Promise<PingResponse> {
  const { data, status } = await ApiRequest.request<PingResponse>(
    '/fee/ping',
    'GET',
    {},
    {}
  );
  if (!status) {
    throw new Error();
  }
  return data;
}
/**
 * 机器节点独占账单信息
 */
export type NodeDistributeBill = {
  id: number;
  applyID: number;
  nodeDistributeID: number;
  fee: number;
  payFee: number;
  payFlag: number;
  payTime: number;
  payType: number;
  payMessage: number;
  userID: number;
  userUsername: string;
  userName: string;
  userGroupID: number;
  createTime: number;
  extraAttributes: string;
};

/**
 * 分页查询机器节点独占账单信息
 */
export async function paginationQueryNodeDistributeBill(
  pageIndex: number,
  pageSize: number
): Promise<PaginationQueryResponse<NodeDistributeBill>> {
  const resp = await ApiRequest.request<
    PaginationQueryResponse<NodeDistributeBill>
  >('/fee/distribute', 'GET', {
    pageSize,
    pageIndex,
  });
  if (!resp.status) {
    throw new Error(resp.message);
  }
  for (const info of resp.data.Data) {
    undefinedWithDefault(info, 'fee', 0.0);
    info.fee.toFixed(2);
    undefinedWithDefault(info, 'payFee', 0);
    info.payFee.toFixed(2);
  }
  return resp.data;
}

/*
 * 节点独占费率
 */
export type NodeDistributeFeeRate = {
  rate36CPU: number;
  rate4GPU: number;
  rate8GPU: number;
};

/**
 * 支付机器独占账单请求参数
 */
export type PayNodeDistributeBillParam = {
  id: number;
  payMoney: number;
  payType: number;
  payMessage: string;
};

/**
 * 支付机器独占账单
 */
export async function payNodeDistributeBill(
  param: PayNodeDistributeBillParam
): Promise<boolean> {
  const resp = await ApiRequest.request(
    '/fee/distribute',
    'PUT',
    {},
    {
      ...param,
    }
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return true;
}

/**
 * 查询机器独占账单费率
 */
export async function queryNodeDistributeFeeRate(): Promise<NodeDistributeFeeRate> {
  const resp = await ApiRequest.request<NodeDistributeFeeRate>(
    '/fee/rate/distribute',
    'GET'
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return resp.data;
}

export type NodeWeekUsageBill = {
  id: number;
  userID: number;
  username: string;
  name: string;
  wallTime: number;
  gwallTime: number;
  fee: number;
  payFee: number;
  startTime: number;
  endTime: number;
  payFlag: number;
  payTime: number;
  payType: number;
  payMessage: string;
  userGroupID: number;
  createTime: number;
  extraAttributes: string;
};

/**
 * 分页查询机器节点时长周账单记录
 */
export async function paginationQueryNodeWeekUsageBill(
  pageIndex: number,
  pageSize: number,
  startDateMilliUnix: number,
  endDateMilliUnix: number
): Promise<PaginationQueryResponse<NodeWeekUsageBill>> {
  const resp = await ApiRequest.request<
    PaginationQueryResponse<NodeWeekUsageBill>
  >('/fee/usage/week', 'GET', {
    pageIndex,
    pageSize,
    startDateMilliUnix,
    endDateMilliUnix,
  });

  if (!resp.status) {
    throw new Error(resp.message);
  }
  for (const single of resp.data.Data) {
    undefinedWithDefault(single, 'wallTime', 0);
    undefinedWithDefault(single, 'gwallTime', 0);
    undefinedWithDefault(single, 'payFlag', 0);
    undefinedWithDefault(single, 'fee', 0);
    undefinedWithDefault(single, 'payType', 0);
  }
  return resp.data;
}

/**
 * 用户组的机器时长账单信息
 */
export type NodeWeekUsageBillForGroup = {
  wallTime: number;
  gwallTime: number;
  fee: number;
  payFee: number;
  userGroupID: number;
};

export async function paginationQueryGroupNodeWeekUsageBill(
  pageIndex: number,
  pageSize: number,
  payFlag: boolean
): Promise<PaginationQueryResponse<NodeWeekUsageBillForGroup>> {
  const resp = await ApiRequest.request<
    PaginationQueryResponse<NodeWeekUsageBillForGroup>
  >('/fee/usage/group/week', 'GET', {
    pageIndex,
    pageSize,
    payFlag,
  });
  if (!resp.status) {
    throw new Error(resp.message);
  }
  for (const info of resp.data.Data) {
    undefinedWithDefault(info, 'wallTime', 0);
    undefinedWithDefault(info, 'gwallTime', 0);
    undefinedWithDefault(info, 'fee', 0);
    undefinedWithDefault(info, 'payFee', 0);
  }
  return resp.data;
}

/**
 * 支付用户组机器节点时长账单参数定义
 */
export type PayGroupNodeUsageBillsParam = {
  userGroupID: number;
  payType: number;
  payMessage: string;
  needFee: number;
};

/**
 * 更新用户组节点使用时长状态(支付)
 */
export async function updateGroupNodeUsageBills(
  param: PayGroupNodeUsageBillsParam
): Promise<number> {
  const resp = await ApiRequest.request<{
    count: number;
  }>(
    '/fee/usage/group/bill',
    'PUT',
    {},
    {
      ...param,
    }
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return resp.data.count;
}

/**
 * 机器节点机时费率
 */
export type nodeUsageFeeRate = {
  cpu: number;
  gpu: number;
};

export async function queryNodeUsageFeeRateInfo(): Promise<nodeUsageFeeRate> {
  const resp = await ApiRequest.request<nodeUsageFeeRate>(
    '/fee/rate/usage',
    'GET'
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return resp.data;
}

/**
 * 机器存储账单
 */
export type NodeQuotaBill = {
  id: number;
  userID: number;
  name: string;
  username: string;
  userGroupID: number;
  operType: number;
  oldSize: number;
  newSize: number;
  oldEndTimeUnix: number;
  newEndTimeUnix: number;
  fee: number;
  payFlag: number;
  payFee: number;
  payTimeUnix: number;
  payType: number;
  payMessage: string;
  createTime: number;
  extraAttributes: string;
};

/**
 * 分页查询机器存储账单
 */
export async function paginationQueryNodeQuotaBill(
  pageIndex: number,
  pageSize: number
): Promise<PaginationQueryResponse<NodeQuotaBill>> {
  const resp = await ApiRequest.request<PaginationQueryResponse<NodeQuotaBill>>(
    '/fee/quota',
    'GET',
    {
      pageIndex,
      pageSize,
    }
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  for (const info of resp.data.Data) {
    undefinedWithDefault(info, 'oldSize', 0);
    undefinedWithDefault(info, 'fee', 0);
    undefinedWithDefault(info, 'payFee', 0);
    undefinedWithDefault(info, 'payFlag', 0);
    undefinedWithDefault(info, 'payType', 0);
    undefinedWithDefault(info, 'operType', 0);
  }
  return resp.data;
}

/**
 * 支付机器节点存储账单参数
 */
export type PayNodeQuotaBillParam = {
  billID: number;
  payType: number;
  payMoney: number;
  payMessage: string;
};

/**
 * 更新机器节点账单的支付信息
 */
export async function updateNodeQuotaBillPayInfo(
  param: PayNodeQuotaBillParam
): Promise<boolean> {
  const resp = await ApiRequest.request(
    '/fee/quota/bill',
    'PUT',
    {},
    {
      ...param,
    }
  );

  if (!resp.status) {
    throw new Error(resp.message);
  }
  return true;
}

/**
 * 机器存储费率
 */
export type NodeQuotaFeeRate = {
  basic: number;
  extra: number;
};

/**
 * 查询机器存储费率信息
 */
export async function queryNodeQuotaFeeRate(): Promise<NodeQuotaFeeRate> {
  const resp = await ApiRequest.request<NodeQuotaFeeRate>(
    '/fee/rate/quota',
    'GET'
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return resp.data;
}

/**
 * 设置机器节点费率参数
 */
export type SetNodeDistributeFeeRateParam = {
  rate36CPU: number;
  rate4GPU: number;
  rate8GPU: number;
};

/**
 * 设置机器节点费率
 */
export async function setNodeDistributeFeeRate(
  param: SetNodeDistributeFeeRateParam
): Promise<boolean> {
  const resp = await ApiRequest.request(
    '/fee/rate/distribute',
    'PUT',
    {},
    {
      ...param,
    }
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return true;
}

/**
 * 设置机时费率参数
 */
export type SetNodeUsageFeeRateParam = {
  cpu: number;
  gpu: number;
};

/**
 * 修改机器机时费率
 */
export async function setNodeUsageFeeRate(
  param: SetNodeUsageFeeRateParam
): Promise<boolean> {
  const resp = await ApiRequest.request(
    '/fee/rate/usage',
    'PUT',
    {},
    {
      ...param,
    }
  );
  if (!resp.status) {
    throw new Error(resp.message);
  }
  return true;
}

/**
 * 设置机器存储费率参数
 */
export type SetNodeQuotaFeeRateParam = {
  basic: number;
  extra: number;
};

/**
 * 设置机器存储费率
 */
export async function setNodeQuotaFeeRate(
  param: SetNodeQuotaFeeRateParam
): Promise<boolean> {
  const resp = await ApiRequest.request(
    '/fee/rate/quota',
    'PUT',
    {},
    {
      ...param,
    }
  );

  if (!resp.status) {
    throw new Error(resp.message);
  }
  return true;
}
